### DISCUSSION TITLE
gitSync wait option is not working

### DISCUSSION THREAD

**Discussion by scobit** [apache/airflow#32277] [Q&A] [answered]
Hello!

Airflow ver: 2.6.2
Chart version: 1.10.0
k8s: 1.23.5

Currently I am testing Airflow on k8s. 
gitSync function working fine with local gitlab repo, with external github repo, with or without authentification.
But one parameter works incorrectly - "wait".
As I understand, this is the time in seconds for scheduler, and it will sync DAG with specified repo every N seconds.
I have changed values from 5 to 120, but scheduller still sync DAG with near 4 minutes delay.

Could you help me to find issue of this behaviour.

**Comment by potiuk** [comment]
Which wait you refer to ? Is it something in the chart? Could not find it in the chart.

Or is it something passed to GitSync ? If it is passed to git-sync then better forum might be to ask in their repository/forums because basically we are **just** running git-sync container. We might of course try to help if you specify exactly what parameter you refer to with reference to the git-sync documentation that you have problem with, but possibly this is not the right forum?

**Reply by scobit** [reply]
"Is it something in the chart?" yes, from chart.
This part of values.yaml:

`# interval between git sync attempts in seconds`
`# high values are more likely to cause DAGs to become out of sync between different components`
`# low values cause more traffic to the remote git repository`   
`wait: 30`
`containerName: git-sync`   
 `uid: 65533`   
    
"Could not find it in the chart." - it should be in official helm chart (values.yaml).
 I got it from this instruction:

https://airflow.apache.org/docs/helm-chart/stable/index.html   


"wait" refer to _helpers.yaml template.

    - name: GIT_SYNC_WAIT
      value: {{ .Values.dags.gitSync.wait | quote }}

**Comment by scobit** [comment]
DOC is here:
https://airflow.apache.org/docs/helm-chart/stable/manage-dags-files.html

"Mounting DAGs using Git-Sync sidecar without Persistence"

  --set dags.persistence.enabled=false \
  --set dags.gitSync.enabled=true

it works in general, but "wait" not change interval of sync in seconds. Always about 4 minutes.

**Reply by potiuk** [reply]
You are likely using different (newer) git-sync version than we support. If you look at the chart latest supported version is registry.k8s.io/git-sync/git-sync:v3.6.3 - and  you are likely using git-sync 4 which is incompatible. GIT_SYNC_WAIT should be fine.

**Reply by scobit** [reply]
I used proper git-sync version. As it was in original repo - 3.6.3.

**Reply by potiuk** [reply]
Is the variable properly set in your resource templates? can you please double check if your `git-sync` container has the proper GIT_SYNC_WAIT variable set with good value there (I expect you have some typo somewhere - maybe you have an extra space somewhere that is not visible. You can use tools like k9s for example to inspect your running container, and check if all the variables are set properly, whether you use git-sync container in the right version.

According to 3.6.3 documentation https://github.com/kubernetes/git-sync/tree/v3.6.3. GIT_SYNC_WAIT is the right way to set the sync time in version 3.6.3. So if there are problems there, you might want to check and open issue in Git Sync repo - because we solely rely on them to work as advertised. Possibly also you can make some experiments and see if latest 3.6* version fixes your problem. And if it does - please report it here, then maybe a good idea to upgrade.

**Reply by potiuk** [reply]
You can also use `helm install --dry-run` and see if the right GIT_SYNC_WAIT is set for you (again there might be some typos in your configuration that are not visible by "looking" and this is a good way of checking if your k8s resources are exactly as you expec them - inspecting if GIT_SYNC_WAIT is properly set as environment for the git-sync container you are looking at is something that might help you to see if you have not made any typos.

**Reply by potiuk** [reply]
Also there is a nice `helm-diff` plugin that allows you to **just** see differences when you modify values, so you can see what effect your changes had on the generated resources.

**Comment by potiuk** [comment]
Please let us know what you found.

**Reply by scobit** [reply]
"helm install --dry-run" - ok, GIT_SYNC_WAIT corresponds to value in values.yaml

"Possibly also you can make some experiments and see if latest 3.6* version fixes" - I have tried a lot of version, 3.6.1 / 3.6.3 / 3.6.8 / 3.6.2 / 3.6.7 etc, no effect.

"helm-diff" - no strange changes or spaces in my file.

Actually I don't think that there is any mistake in my actions, but of course we should check.
So I made below sequence of actions. And issue still actual.
As you can see, i have changed only ingress options and address of my github repo.


```
$ helm delete airflow

$ kubectl delete ns airflow

$ helm repo remove apache-airflow

$ helm repo add apache-airflow https://airflow.apache.org
"apache-airflow" has been added to your repositories

$ helm repo update
Hang tight while we grab the latest from your chart repositories...
...
...Successfully got an update from the "apache-airflow" chart repository
...
Update Complete. ⎈Happy Helming!⎈


$ helm show values apache-airflow/airflow > values-original.yaml

$ helm show values apache-airflow/airflow > values.yaml

$ vim values.yaml 


$ diff values.yaml values-original.yaml 
127c127
<     enabled: yes
---
>     enabled: false
142,143c142,143
<     hosts: 
<      - name: "airflow.MyAwesomeDomain.example.com"
---
>     hosts: []
>     # - name: ""
2139c2139
<     enabled: true
---
>     enabled: false
2144,2145c2144,2145
<     repo: https://github.com/MyAwesomeAccount/Some.git
<     branch: main
---
>     repo: https://github.com/apache/airflow.git
>     branch: v2-2-stable
2152c2152
<     subPath: ""
---
>     subPath: "tests/dags"


$ helm upgrade --install airflow apache-airflow/airflow -f values.yaml --namespace airflow --create-namespace
```

**Reply by potiuk** [reply]
Then you should likely raise it to git-sync - show logs of your git-sync where it checks with different intervals, show the env vars you are running it with it and maybe they will be able to help

**Reply by potiuk** [reply]
Again - what we do here, is we run git-sync. If it does not work as advertised and you have alll the reasons to beleive you passed all the right values, just talk to them - I can at most do the same, and having man-in-the-middle to pass your problem to them and discuss it back/forth makes very little sense.

**Reply by scobit** [reply]
ok, i will try to discuss this issue with git-sync team.
But I suppose, that problem not there because:
git-sync is just serving main (airflow-web, scheduller) containers, git-sync has only startup logs. Maximum what we can troubleshoot here - change image versions.

Right now I am testing user-community helm with airflow:

https://github.com/airflow-helm/charts/tree/main

And sync time option working fine there.

Well, I'm afraid that they will probably recommend to use another helm of airflow.

**Reply by potiuk** [reply]
That's fine. You are free to choose and recommend the other chart. We don't force anyone to use ours. 

I just wonder - if you cannot debug it, what does it mean that your: 

```
I have changed values from 5 to 120, but scheduller still sync DAG with near 4 minutes delay.
```

How do you know it if you can't see git-sync working ? Maybe you simply misuderstood the problem and all the time the problem is something completely different ?

**Reply by scobit** [reply]
Sorry for delay with answer, I was out of office for several days.
Of course I can use any helm, but if it possible I would prefer to use Apache community helm.
"How do you know it if you can't see git-sync working ? Maybe you simply misuderstood the problem and all the time the problem is something completely different ?" - actually I don't understand question, please clarify if needed. Above in discussion I mentioned that in general sync is working. DAGs which I put in gitlab or github are displayed in airflow. Not wotking sync time between git repo and airflow. No matter what number of seconds I will specify in "wait" option it will always be about 4 minutes. So when I put new DAG to git repo, I am waiting for 4 minutes when it will appear in airflow.

**Reply by potiuk** [reply, marked as answer]
Then yes. You probably misunderstood completely what git-sync command's `wait` does.

Git sync - synchronizes the files to the local folder. That's it. The fact that git-sync synchronizes the files to the folder, does not mean they will appear in the UI. 

Then you have a scheduler (actually DAGFileProcesoor) that parses the DAG folder regularly - and only after it parses the DAG files there appear in the UI. Those two processes are completely independent. Parsing of the DAGs is asynchronous and generally you have various configuration options to configure how often it parses the files.

The way you can configure it, and trade-offs they come with are explained in https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/scheduler.html#fine-tuning-your-scheduler-performance

So your problem is likely not the "git-sync" wait not working, but that you need to fine tune the scheduler to behave in the way you want (so for example use more CPU while decreasing delays of DAGs appearing in the UI).

**Reply by scobit** [reply]
After some tests, I found required parameter based on scheduler performance doc:
`                            
extraEnv: |                            
  - name: AIRFLOW__SCHEDULER__DAG_DIR_LIST_INTERVAL                                                                       
    value: '10'                        
`

Thank you!!!
